design/src/rfcs/rfc0023_sdk_credential_cache_type_safety.md

```diff
@@ -0,0 +1,260 @@
+    .credentials(ImdsCredentialsProvider::new())
+    .load()
+    .await;
+```
+
+Another method could be added to `ConfigLoader` that makes it easier to use the default cache:
+
+```rust
+let sdk_config = aws_config::from_env()
+    .credentials_with_default_cache(ImdsCredentialsProvider::new())
+    .load()
+    .await;
+```
+
+### Pros/cons
+
+- :+1: It's flexible, and somewhat enforces correct cache setup through types.
+- :+1: Removes the possibility of double caching since the cache implementations won't
+  implement `ProvideCredentials`.
+- :-1: Customers may unintentionally implement `ProvideCachedCredentials` instead of `ProvideCredentials`
+  for a custom provider, and then not realize they're not benefiting from caching.
+- :-1: The documentation needs to make it very clear what the differences are between `ProvideCredentials`
+  and `ProvideCachedCredentials` since they will look identical.
+- :-1: It's possible to implement both `ProvideCachedCredentials` and `ProvideCredentials`, which
+  breaks the type safety goals.
+
+Option B: `CacheCredentials` trait
+----------------------------------
+
+This option is similar to Option A, except that the cache trait is distinct from `ProvideCredentials` so
+that it's more apparent when mistakenly implementing the wrong trait for a custom credentials provider.
+
+A `CacheCredentials` trait would be added that looks as follows:
+```rust
+pub trait CacheCredentials: Send + Sync + Debug {
+    pub async fn cached(&self, now: SystemTime) -> Result<Credentials, CredentialsError>;
+}
+```
+
+Instances implementing `CacheCredentials` need to own the `ProvideCredentials` implementation
+to make both lazy and eager credentials caching possible.
+
+The configuration examples look identical to Option A.
+
+### Pros/cons
+
+- :+1: It's flexible, and enforces correct cache setup through types slightly better than Option A.
+- :+1: Removes the possibility of double caching since the cache implementations won't
+  implement `ProvideCredentials`.
+- :-1: Customers can still unintentionally implement the wrong trait and miss out on caching when
+  creating custom credentials providers, but it will be more apparent than in Option A.
+- :-1: It's possible to implement both `CacheCredentials` and `ProvideCredentials`, which
+  breaks the type safety goals.
+
+Option C: `CredentialsCache` struct
+-----------------------------------
+
+The struct approach posits that customers don't need or want to implement custom credential caching,
+but at the same time, doesn't make it impossible to add custom caching later.
+
+The idea is that there would be a struct called `CredentialsCache` that specifies the desired
+caching approach for a given credentials provider:
+
+```rust
+pub struct LazyCache {
+    credentials_provider: Arc<dyn ProvideCredentials>,
+    // ...
+}
+
+pub struct EagerCache {
+    credentials_provider: Arc<dyn ProvideCredentials>,
+    // ...
+}
```

Comment:
I like this design, but I'm struck by the fact that we might be composing unnecessarily–what if we separate the cache from the inner credentials provider in this part of the design?

You would then be able to construct a "cache strategy" separate from the credentials provider. When the default chain was actually loaded, you could assemble the concrete impl. 
This also happens to address the fact that under the hood, an eager the cache and a lazy cache are the same code, we just pre-call provide credentials on the eager cache

Here's what I'm thinking:

```rust
/// Signal-struct to denote the cache mode (no logic)
pub struct CredentialsCache {
  inner: CredentialsCacheInner
}
enum CredentialsCacheInner {
  Lazy,
  Eager,
  NoCaching
}
impl CredentialsCache {
  pub fn lazy() -> Self { .. }
  pub fn eager() -> Self { .. }
  pub(crate) fn instantiate(provider: Box<dyn ProvideCredentials>, conf: ProviderConfig) -> ConcreteCredentialsCache { .. // private function used to actually make the chain }
}

async fn main() {
  // custom provider, default cache
  aws_config::load_from_env().credentials(ImdsProvider::new()).load(); // uses default cache
  // default provide , but no caching
  aws_config::load_from_env()
    .credentials_cache(CredentialsCache::no_cache())
    .load(); 

```

Reply:
I really like how this removes the need to think about caching when changing the credentials provider while still resolving the core issue. It's only slightly different from Option C, so I've opted to set it as the selected design that the doc focuses on, and move the rest of the alternatives to an appendix.